core/types/transaction.go

```diff
@@ -415,9 +415,7 @@ func (t *TransactionsByPriceAndNonce) Peek() *Transaction {
 }
 
 // Shift replaces the current best head with the next one from the same account.
-func (t *TransactionsByPriceAndNonce) Shift() {
-	signer := deriveSigner(t.heads[0].data.V)
-	// derive signer but don't cache.
+func (t *TransactionsByPriceAndNonce) Shift(signer Signer) {
```

Comment:
@fjl Thanks for the feedback. I updated my commit to store the signer in `TransactionsByPriceAndNonce` and check the sender address before we construct it. Please help review. Thanks!